core/src/lib.rs

```diff
@@ -73,34 +78,78 @@ pub type DefaultFileService = FileServiceImpl<
     DefaultFileEncryptionService,
 >;
 
-static FAILURE_DB: &str = "FAILURE<DB_ERROR>";
-static FAILURE_ACCOUNT: &str = "FAILURE<ACCOUNT_MISSING>";
-static FAILURE_META_CREATE: &str = "FAILURE<META_CREATE>";
-static FAILURE_FILE_GET: &str = "FAILURE<FILE_GET>";
-static FAILURE_ROOT_GET: &str = "FAILURE<ROOT_GET>";
-static FAILURE_UUID_UNWRAP: &str = "FAILURE<UUID_UNWRAP>";
+#[repr(C)]
+pub struct ResultWrapper {
+    is_error: bool,
+    value: Value,
+    error: LockbookError,
+}
 
-unsafe fn string_from_ptr(c_path: *const c_char) -> String {
-    CStr::from_ptr(c_path)
-        .to_str()
-        .expect("Could not C String -> Rust String")
-        .to_string()
+#[repr(C)]
+pub union Value {
+    success: *const c_char,
+    error: *const c_char,
 }
 
-unsafe fn connect_db(c_path: *const c_char) -> Option<Db> {
-    let path = string_from_ptr(c_path);
-    let config = Config {
-        writeable_path: path,
-    };
-    match DefaultDbProvider::connect_to_db(&config) {
-        Ok(db) => Some(db),
-        Err(err) => {
-            error!("DB connection failed! Error: {:?}", err); // TEMP HERE
-            None
+// TODO: make these better (they're the errors clients will actually handle)
+#[repr(C)]
+pub enum LockbookError {
+    Network,
+    Database,
+}
+
+impl<T: Serialize, E: Debug> From<Result<T, E>> for ResultWrapper {
+    fn from(result: Result<T, E>) -> Self {
+        ResultWrapper {
+            is_error: result.is_err(),
+            value: {
+                match result {
+                    Ok(value) => Value {
+                        success: CString::new(json!(value).to_string()).unwrap().into_raw(),
+                    },
+                    Err(err) => Value {
+                        error: CString::new(format!("{:?}", err)).unwrap().into_raw(),
+                    },
+                }
+            },
+            error: LockbookError::Database,
```

Comment:
Is this supposed to be a match or something?

Reply:
No not yet that was just me making an enum